[BEAM-160] Port 'NexMark Queries' to Beam for use as integration test

echauchot opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`.
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
R: @dpmills @dhalperi
CC: @stasl @aviemzur @aljoscha  because we discussed NexMark together :)
CC: @mshields822 I know you don not work on it anymore, but you might be interested :)
CC: @ssisk for reflexion for IT tests

This is a port of the NexMark queries to Beam, to be used as integration tests.
This can also be used as A-B testing (no-regression or performance comparison between 2 versions of the same engine or of the same runner)

This a continuation of the previous PRs (https://github.com/apache/beam/pull/99 and https://github.com/apache/beam/pull/366) from Mark Shields.
The code has changed quite a bit: some queries have changed to use new Beam APIs and there where some big refactorings. More important, we can now run all the queries in all the runners. Nevertheless, there are still some open issues in Nexmark (https://github.com/iemejia/beam/issues) and in Beam upstream (see issue links in https://issues.apache.org/jira/browse/BEAM-160)

Here is a doc that present NexMark components and pseudo code of the queries to ease the review : https://drive.google.com/open?id=1VgnGiVu8vSfm7Et-xAtQYv0PlEpqeyfmhpQUNPmWRJs 

Everything needed to launch the queries is in the Readme. There is also a support matrix towards the runners.

Please do not squash commits because there are several authors Mark, Ismaël and I.

Good review :) !

jbonofre commented:
R: @jbonofre

dhalperi commented:
R: -@dhalperi

dpmills commented on integration/java/nexmark/src/main/java/org/apache/beam/integration/nexmark/NexmarkLauncher.java:
```diff
@@ -0,0 +1,1172 @@
+    return options.isStreaming();
+  }
+
+  /**
```
Why are these hard-coded?

echauchot replied:
This option allows the user to specify if we wants to run the pipelines in streaming mode or in batch mode. If he sets `streaming=true` in the options, then the first step of the pipelines will be 
`Read.from(new UnboundedEventSource)` otherwise it will be `Read.from(new BoundedEventSource)`

echauchot replied:
IMHO I think they are leftovers: `coresPerWorker` is never used, so I will remove it. And `maxNumWorkers()` is only used in query 10 (write event logs to Google Cloud Storage) to define the number of shards. 

 I will start a discussion on the ML about whether we should keep query 10 like this (with Google specifics as the aim was to test a specific Google customer use case) or use a more generic (`Beam filesystems`) data storage technology. If so, the configuration parameter is likely to disappear also. But this refactoring of query 10 does not block the PR I think. It could be done afterwards.

dpmills commented on integration/java/nexmark/src/main/java/org/apache/beam/integration/nexmark/NexmarkLauncher.java:
```diff
@@ -0,0 +1,1172 @@
+   */
+  private void invokeBuilderForPublishOnlyPipeline(PipelineBuilder<NexmarkOptions> builder) {
+    builder.build(options);
+//    throw new UnsupportedOperationException(
```
Delete commented code if it's not needed

echauchot replied:
Yes .The PR in general lacks  a big cleaning :)

dpmills commented on integration/java/nexmark/src/main/java/org/apache/beam/integration/nexmark/NexmarkLauncher.java:
```diff
@@ -0,0 +1,1172 @@
+   * If monitoring, wait until the publisher pipeline has run long enough to establish
+   * a backlog on the Pubsub topic. Otherwise, return immediately.
+   */
+  private void waitForPublisherPreload() {
```
Consider just deleting this along with the other preload-related functionality, since it doesn't work in this PR

echauchot replied:
+1

echauchot replied:
I only left `conf.preloadSeconds` that is used to count initialization time in the time before cancelling the job.

dpmills commented on integration/java/nexmark/src/main/java/org/apache/beam/integration/nexmark/NexmarkLauncher.java:
```diff
@@ -0,0 +1,1172 @@
+          sinkEventsToAvro(source);
+        }
+
+        // Special hacks for Query 10 (big logger).
```
Add more comments about what the hacks are and why they are needed

echauchot replied:
ok

dpmills commented on integration/java/nexmark/src/main/java/org/apache/beam/integration/nexmark/NexmarkUtils.java:
```diff
@@ -0,0 +1,655 @@
+   * Setup pipeline with codes and some other options.
+   */
+  public static void setupPipeline(CoderStrategy coderStrategy, Pipeline p) {
+    //TODO Ismael check
```
The new direct runner doesn't support disabling randomization at all, so this can be deleted.  Do the tests pass on DirectRunner without it?

echauchot replied:
Yes, all the queries and and all the unit tests pass in direct runner in batch and in streaming mode even if randomization is not disabled so I can safely remove this piece of code.

dpmills commented on integration/java/nexmark/src/main/java/org/apache/beam/integration/nexmark/NexmarkUtils.java:
```diff
@@ -0,0 +1,655 @@
+  /**
+   * Return a transform to write given number of bytes to durable store on every record.
+   */
+  public static <T> ParDo.SingleOutput<T, T> diskBusy(String name, final long bytes) {
```
This can use the State API now

echauchot replied:
Yes, forgot that one, thanks!

dpmills commented on integration/java/nexmark/src/main/java/org/apache/beam/integration/nexmark/model/Event.java:
```diff
@@ -0,0 +1,166 @@
+    public void encode(Event value, OutputStream outStream)
+        throws CoderException, IOException {
+      if (value.newPerson != null) {
+        INT_CODER.encode(0, outStream);
```
Put the tag values in an enum

echauchot replied:
Missed that one also, thanks!

echauchot commented:
@dpmills thanks for your comments, I addressed all of them, PTAL.

dpmills commented:
LGTM

I'm not a committer however, so over to @dhalperi for comitting.

dhalperi commented:
I don't know how to merge this. Too many commits, too many authors, request not to squash.

This is going to really pollute history if merged as-is, but that is quite possibly by design.

I'll let @iemejia do it, he's a committer :)

dhalperi commented:
Additionally, I believe there are as-yet-unresolved issues about module names and paths of code.

R: @davorbonaci  for some input.

iemejia commented:
@dhalperi I will see with @echauchot to reduce the number of commits, I think we can still squash some of the latest commits (that addressed the review comments + did the latest fixes before the PR), anyway you have to consider that this is a long ongoing work with three different authors so it is normal to have a reasonable number of commits, I won't consider this pollution because there are not intentionally tiny commits, most of them are like this by design, but we will see still how to reduce them.

iemejia commented:
For reference @davorbonaci the last point Dan mentions is the paths, currently Nexmark uses the original path that Mark had for the PR: integration/java/nexmark This introduces a new directory structure for integration tests on Beam, this can make sense I think and even be used to have also part of the things @ssisk is working on.
In a brief discussion with Dan we also talked about the possibility of putting this into the examples directory but after discussing with @echauchot we considered that it didn't make sense either because Nexmark is not just a example, it has clearly a different scope.
So in the end we need you to confirm us if you are ok with this new path or to propose us some alternative so we can refactor it if needed.

echauchot commented:
@dhalperi we managed to reduce the number of commits from 55 to 27: we wanted to keep some big refactorings or some big subjects (like queries fixing/improving) separated. I just pushed the branch. WDYT?

davorbonaci commented on integration/java/nexmark/BEAM_ON_FLINK_ON_GCP.md:
```diff
@@ -0,0 +1,282 @@
+<!--
```
sdks/java/nexmark?

echauchot replied:
It would be great to have it in the sdk! I am in favor of it. The only question I ask myself is: Nexmark is more a benchmarking suite, not an API or framework. Is it not a problem to have a piece of software that is not used to build software in the sdk module?

davorbonaci commented on integration/java/nexmark/BEAM_ON_FLINK_ON_GCP.md:
```diff
@@ -0,0 +1,282 @@
+    under the License.
+-->
+
+# Running NexMark on Beam on Flink on Google Compute Platform
```
Move to the website; contribution section?

echauchot replied:
Thanks for pointing that out! Actually it is a leftover, I forgot that file. I propose to simply remove it because we have not tested to run Nexmark that way since Mark did; more important, I prefer communicating about running NexMark on a standard Flink cluster similarly to what is described in the readme about Spark cluster.

davorbonaci commented on integration/java/nexmark/README.md:
```diff
@@ -0,0 +1,340 @@
+    under the License.
+-->
+
+# NEXMark integration suite
```
website?

echauchot replied:
I guess you mean having this content in both beam website and the readme, right?

echauchot replied:
Ok, what I propose is to add a short introduction to Nexmark similar to the talk we gave at the ApacheCon and just reference the `Readme` (once it is merged) to avoid duplication of information.

davorbonaci commented on integration/java/nexmark/pom.xml:
```diff
@@ -0,0 +1,292 @@
+    <relativePath>../pom.xml</relativePath>
+  </parent>
+
+  <artifactId>beam-integration-java-nexmark</artifactId>
```
beam-sdks-java-nexmark

davorbonaci commented on integration/java/nexmark/pom.xml:
```diff
@@ -0,0 +1,292 @@
+
+  <parent>
+    <groupId>org.apache.beam</groupId>
+    <artifactId>beam-integration-java-parent</artifactId>
```
beam-sdks-java-parent

echauchot replied:
+1 modulo my interrogation in your first comment

davorbonaci commented on integration/java/nexmark/pom.xml:
```diff
@@ -0,0 +1,292 @@
+  </parent>
+
+  <artifactId>beam-integration-java-nexmark</artifactId>
+  <name>Apache Beam :: Integration Tests :: Java :: Nexmark</name>
```
name update?

echauchot replied:
You mean renaming to `Apache Beam :: SDKs :: Java :: Nexmark` or renaming NexMark itself to something  like `Nexmark Test Suite` ?

davorbonaci commented on integration/java/nexmark/pom.xml:
```diff
@@ -0,0 +1,292 @@
+      </dependencies>
+    </profile>
+
+    <!-- Include the Google Cloud Dataflow runner -P dataflow-runner -->
```
rm comment?

davorbonaci commented on integration/java/nexmark/pom.xml:
```diff
@@ -0,0 +1,292 @@
+      </dependencies>
+    </profile>
+
+    <!-- Include the Apache Spark runner -P spark-runner -->
```
rm?

davorbonaci commented on integration/java/nexmark/pom.xml:
```diff
@@ -0,0 +1,292 @@
+      </dependencies>
+    </profile>
+
+    <!-- Include the Apache Flink runner with -P flink-runner -->
```
rm?

davorbonaci commented on integration/java/nexmark/pom.xml:
```diff
@@ -0,0 +1,292 @@
+      </dependencies>
+    </profile>
+
+    <!-- Include the Apache Apex runner with -P apex-runner -->
```
rm?

davorbonaci commented on integration/java/nexmark/pom.xml:
```diff
@@ -0,0 +1,292 @@
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-shade-plugin</artifactId>
```
possibly in a runner-specific profile? see examples & archetypes for more patterns around this.

echauchot replied:
or I can remove it, the plugin configuration should be inherited from the parent pom. Thanks

echauchot replied:
Actually, shade plugin is configured differently in parent, so I will leave the configuration in nexmark module. 
For Runner specific dependencies, they are not included in the bundled jar because they are scoped runtime, so I'm not sure I should configure shade plugin in runner-specific profiles.

davorbonaci commented on integration/java/nexmark/pom.xml:
```diff
@@ -0,0 +1,292 @@
+        </executions>
+      </plugin>
+
+      <!-- Coverage analysis for unit tests. -->
```
unnecessary? should be inherited, no?

echauchot replied:
Yes, thanks

davorbonaci commented on integration/java/nexmark/pom.xml:
```diff
@@ -0,0 +1,292 @@
+  </build>
+
+  <dependencies>
+    <!-- Java SDK -->
```
rm all these comments, this is obvious.

davorbonaci commented on integration/java/nexmark/src/main/java/org/apache/beam/integration/nexmark/package-info.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ * Nexmark.
+ */
+package org.apache.beam.integration.nexmark;
```
Please double-check the output of the aggregated javadoc. Nexmark packages shouldn't be there.

echauchot replied:
ok

davorbonaci commented on integration/java/nexmark/src/main/java/org/apache/beam/integration/nexmark/queries/Query0.java:
```diff
@@ -0,0 +1,71 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.integration.nexmark.queries;
```
org.apache.beam.sdk.nexmark.XXX (across the board)

davorbonaci commented on pom.xml:
```diff
@@ -159,6 +159,7 @@
     <module>sdks</module>
     <module>runners</module>
     <module>examples</module>
+    <module>integration</module>
```
revert?

davorbonaci commented on integration/pom.xml:
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
revert?

davorbonaci commented on integration/java/pom.xml:
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
revert?

echauchot commented:
Thanks @davorbonaci for your comments ! I'll be in meetings today, but I'll address them ASAP.

echauchot commented:
Davor, thanks for your review, I have addressed your comments, see the last 4 commits. I have also rebased on master. I will squash the last 4 commits into the 5th when you find the PR is ok.

reuvenlax commented:
Is this PR blocked on something right now?

iemejia commented:
Not really, we are just waiting for @davorbonaci to give his LGTM for the changes he requested and then it will be ready to go. I really want to see this merged. Btw Davor can you take a final look so we merge this.

Note: After the LGTM we will squash the extra commits of the review + pump the version to 2.2.0-SNAPSHOT.

echauchot commented:
@jbonofre moved to 2.2.0-SNAPSHOT and rebased on master

reuvenlax commented:
@davorbonaci, any more concerns with this PR?

echauchot commented:
I'm very glad to see nexmark move forward, thanks! I did a final squash of the review comments.

mshields822 commented:
Thanks for all your great work on this Etienne and Ismaël! Happy to see it closed at last.

Best,
-Mark

iemejia commented:
No, the big credit is up to you @mshields822 this would not have been possible at all without your work, we just pushed it a bit to the last line, you did the long journey. Thanks for the support and of course for your nice creation.

echauchot commented:
I could not have said it better :)
